File: packages/games_api/lib/src/models/user_achievement.dart
import 'achievement.dart';

/// Represents an unlocked achievement for a user
class UserAchievement {
  final int id;
  final int userId;
  final int achievementId;
  final DateTime? unlockedAt;
  final String progress;
  final DateTime createdAt;
  final DateTime updatedAt;
  final Achievement? achievement;

  UserAchievement({
    required this.id,
    required this.userId,
    required this.achievementId,
    this.unlockedAt,
    required this.progress,
    required this.createdAt,
    required this.updatedAt,
    this.achievement,
  });

  factory UserAchievement.fromJson(Map<String, dynamic> json) {
    return UserAchievement(
      id: json['id'] ?? 0,
      userId: json['user_id'] ?? 0,
      achievementId: json['achievement_id'] ?? 0,
      unlockedAt: json['unlocked_at'] != null ? DateTime.parse(json['unlocked_at']) : null,
      progress: json['progress'] ?? '{}',
      createdAt: DateTime.parse(json['created_at'] ?? DateTime.now().toIso8601String()),
      updatedAt: DateTime.parse(json['updated_at'] ?? DateTime.now().toIso8601String()),
      achievement: json['achievement'] != null ? Achievement.fromJson(json['achievement']) : null,
    );
  }
}

File: api/app/models/user_achievement.go
package models

import (
	"base/core/app/profile"
	"time"

	"gorm.io/gorm"
)

// UserAchievement tracks which achievements users have unlocked
type UserAchievement struct {
	Id            uint           `gorm:"column:id;primary_key;auto_increment" json:"id"`
	UserId        uint           `gorm:"column:user_id;not null;index" json:"user_id" validate:"required"`
	User          *profile.User  `json:"user,omitempty" gorm:"foreignKey:UserId"`
	AchievementId uint           `gorm:"column:achievement_id;not null;index" json:"achievement_id" validate:"required"`
	Achievement   *Achievement   `json:"achievement,omitempty" gorm:"foreignKey:AchievementId"`
	Progress      string         `gorm:"column:progress;type:json" json:"progress"` // JSON for partial completion tracking
	UnlockedAt    *time.Time     `gorm:"column:unlocked_at" json:"unlocked_at"`
	CreatedAt     time.Time      `gorm:"column:created_at" json:"created_at"`
	UpdatedAt     time.Time      `gorm:"column:updated_at" json:"updated_at"`
	DeletedAt     gorm.DeletedAt `gorm:"column:deleted_at;index" json:"-"`
}

func (UserAchievement) TableName() string {
	return "user_achievements"
}
